### ISSUE TITLE
nsqd: /stats is slow for large deployments

### ISSUE THREAD

**Issue by jinhailang** [nsqio/nsq#920]
hello:

I have a question. When too many clients (less than a thousand) to connect to the NSQ service, nsqadmin web will slowly (It may take more than a minute), and may timeout failed.

![image 1](https://user-images.githubusercontent.com/5457063/28672272-e17b2c72-7311-11e7-9215-f9e4865000c8.png)

When NSQD processing "/stats" request , the CPU will rise suddenly：

normal：
<img width="589" alt="2017-07-27 20 50 26" src="https://user-images.githubusercontent.com/5457063/28671139-2c065c34-730e-11e7-983b-851a8ddcaa67.png">

get "/stats?format=json":
<img width="620" alt="2017-07-27 20 52 58" src="https://user-images.githubusercontent.com/5457063/28671178-506e79d0-730e-11e7-8a66-55c50ee83090.png">

**I read the nsqd code, the [`GetStats`](https://github.com/nsqio/nsq/blob/6c64ce0d958a844106715cfaf52d7bd8ceef4af6/nsqd/stats.go#L116) function will query all topics and channels. Perhaps only query the specified topicsc or channel efficiency will be much higher.**

**Can you give me some advice? thank you very much！**

I have four nsqd, and four nsqlookup.
```verbose = false
tcp_address = "0.0.0.0:4150"
http_address = "0.0.0.0:4151"
nsqlookupd_tcp_addresses = [
    "192.168.12.15:4160",
    "192.168.12.16:4160",
    "192.168.12.17:4160",
    "192.168.12.18:4160"
]
broadcast_address = "cdn-nsqd1.pxy.domain.com"
data_path = "/disk/ssd1/nsqdata"
http_client_connect_timeout = "10s"
http_client_request_timeout = "30s"
mem_queue_size = 10000
max_bytes_per_file = 104857600
sync_every = 2500
sync_timeout = "2s"
msg_timeout = "60s"
max_msg_timeout = "15m"
max_msg_size = 1024768
max_req_timeout = "1h"
max_body_size = 5123840
max_heartbeat_interval = "60s"
max_rdy_count = 2500
max_output_buffer_size = 65536
max_output_buffer_timeout = "1s"
statsd_prefix = "nsq.%s"
statsd_interval = "60s"
statsd_mem_stats = true
e2e_processing_latency_percentiles = [
    100.0,
    99.0,
    95.0
]
e2e_processing_latency_window_time = "10m"
tls_required = true
tls_min_version = ""
tls_cert = "/usr/local/nsq/etc/certs/ca.crt"
tls_key = "/usr/local/nsq/etc/certs/ca.key"
tls_root_ca_file = "/usr/local/nsq/etc/certs/ca.crt"
tls_client_auth_policy = "require-verify"
deflate = true
max_deflate_level = 6
snappy = true
```

**Comment by ploxiln** [comment]
Roughly  how many topics, channels, and consumers per channel do you have? What version of nsqd/nsqadmin are you running, 1.0.0-compat?

**Comment by jinhailang** [comment]
I have 4 topics, about 100 channels, Each channel has 4 to 5 consumers.

<img width="1211" alt="2017-07-28 09 48 44" src="https://user-images.githubusercontent.com/5457063/28699141-700ae4d8-737a-11e7-81bf-efc012ac0fbc.png">

**Comment by girishramnani** [comment]
Hi,
I would like to contribute to this.
One of the possible solution can be to allow the user to pass a query parameter "topic" which takes
a regex as a value, eg: "/stats?topic=hello*" will show stats for topics which start with "hello". 

Any suggestions?

**Comment by jinhailang** [comment]
API has been allowed the user to pass a query parameter "topic" which takes a value, but `GetStats` function is not used.
https://github.com/nsqio/nsq/blob/6c64ce0d958a844106715cfaf52d7bd8ceef4af6/nsqd/http.go#L470

**Comment by girishramnani** [comment]
So here, the stats for all topics are collected and then the topic if given is filtered. Just as an optimization what if stats of only the specified topics are fetched. Would that help?

**Comment by jinhailang** [comment]
Yes, I think it would be much faster,.

**Comment by girishramnani** [comment]
ok, I will work on that then.

**Comment by mreiferson** [comment]
At some point we improved the locking semantics of getstats, but that wouldn't necessarily improve the raw performance of large deployments.

If `nsqadmin` is querying inefficiently, +1 to improving that @girishramnani.

**Comment by ploxiln** [comment]
So there's two things:

  * nsqadmin does not use the "topic=?" param when getting stats for a particular topic
  * nsqd gets stats for all topics, then if "topic" was specified trims to only that topic

Just having nsqadmin use the "topic" param could help a noticeable amount - it's possible nsq spends a lot of time serializing json and sending it, for huge stats responses. That work is saved by specifying "topic", even if the stats gathering work inside nsqd is not (currently) saved.

**Comment by girishramnani** [comment]
Yeah,. I will submit a PR by this weekend.

**Comment by jinhailang** [comment]
@ploxiln @girishramnani Thank you！I tried to change it.

PR: https://github.com/nsqio/nsq/pull/925

However, when the request is large, the client is connected too much, gets stats is still too slow.
I think it is `getstats` caused by the use of the `lock`, the use of `chan` will be better.

**Comment by ploxiln** [comment]
this should be at least partially fixed by #929

**Comment by mreiferson** [comment]
Work discussed here completed in #929
